### ISSUE TITLE
RFC Shift schema PHP api out of LeftAndMain

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-admin#348]
# Introduction

At the moment the php form schema API is hard-coded into LeftAndMain with the following endpoints:

 - methodSchema()
 - schema()
 - getFormSchema()
 - setFormSchema()
 - getSchemaResponse()
 
# Recommendation
 
Instead, I suggest to shift this into LeftAndMainFormRequestHandler. The `schema` method would be moved into this request handler as a default `allowed_action`.

# Rationale

Primarily this would flatten and simplify the role of LeftAndMain.

The other aim is to eliminate unnecessary complexity required in routing schema urls. For instance, we need separate schema / methodSchema to resolve the `schema` action for any forms, and potentially more methods would need to be added for different form URLS.

Shifting this into a `schema` action on the request handler would automatically provide a schema route for any pre-existing form route.

For instance, the schema url

`/admin/methodSchema/Modals/remoteCreateForm`

would become:

`/admin/Modals/remoteCreateForm/schema`

**Comment by flamerohr** [comment]
Also include moving FormField schema specific logic, we can achieve this with the use of extensions and schema data could be handled by generators.

@tractorcow described it to me much more succinctly than I can put into words.

**Comment by robbieaverill** [comment]
Would make https://github.com/silverstripe/silverstripe-admin/issues/240 obsolete

**Comment by flamerohr** [comment]
Yes, the whole idea of moving all the schema out is so we could pull it out of framework entirely when it's not needed (e.g. we built a new admin which doesn't use the schema, or a framework only install)

**Comment by tractorcow** [comment]
Details I discussed with @flamerohr:

 - formfields dont manipulate schema anymore. There's a generator class (injected by extension) that does that
 - yaml can inject specific subclasses of generator for specific types
 - shift LeftAndMainFormRequestHandler to this module. Make the name something smarter. Like SchemaFormRequestHandler.
 - admin module requires this new module

**Comment by flamerohr** [comment]
I think `LeftAndMainFormRequestHandler` is very handy in itself (field actions), and it doesn't need to be coupled with schema logic.

But `SchemaFormRequestHandler` could be extending it!

**Comment by tractorcow** [comment]
All LeftAndMainFormRequestHandler does is add another URL segment. that could be moved to core FormRequestHandler though.

**Comment by flamerohr** [comment]
True, good point
